### PR TITLE
feature: send voice message API

### PR DESCRIPTION
**Pull request by alfogrillo**
This PR adds a ffi binding for sending voice messages in the legacy format (before extensible events). It also makes minor changes in the `matrix-sdk` crate to accomodate this change.

### CODE REVIEW

File: crates/matrix-sdk/src/encryption/mod.rs [hunk 1]
```diff
@@ -1352,4 +1364,4 @@ mod tests {
         let after_taking_lock_second_time = client.olm_machine().await.as_ref().unwrap().clone();
         assert!(after_taking_lock_first_time.same_as(&after_taking_lock_second_time));
     }
-}
+}
```
Comment (jplatte) [inline crates/matrix-sdk/src/encryption/mod.rs:1367]: You removed the final newline in this file.

**Review by jplatte — COMMENTED** [review]
You can undo the latest commit now and run `cargo update -p ruma-events` instead to get a crates.io version that includes the new constructors.

File: crates/matrix-sdk/src/encryption/mod.rs (hunk 1 above)
Comment (alfogrillo) [reply crates/matrix-sdk/src/encryption/mod.rs]: Fixed [here](https://github.com/matrix-org/matrix-rust-sdk/pull/2697/commits/5b1178d3b638095c7c5c27ca6b6d07edec87d455)

**Comment by alfogrillo** [comment]
> You can undo the latest commit now and run `cargo update -p ruma-events` instead to get a crates.io version that includes the new constructors.

Thanks. 
Done [here](https://github.com/matrix-org/matrix-rust-sdk/pull/2697/commits/5b1178d3b638095c7c5c27ca6b6d07edec87d455).

File: Cargo.lock [hunk 2]
```diff
@@ -4911,9 +4911,9 @@ dependencies = [
 
 [[package]]
 name = "ruma-events"
-version = "0.27.1"
+version = "0.27.2"
```
[the remark this answers is not in the document]
Comment (jplatte) [reply Cargo.lock]: We don't have a direct dependency on `ruma-events`, that's why no `Cargo.toml` is updated.

File: crates/matrix-sdk/src/encryption/mod.rs [hunk 3]
```diff
@@ -202,13 +203,23 @@ impl Client {
...
-                    info: Some(Box::new(info))
+
+                let mut audio_message_event_content = assign!(AudioMessageEventContent::encrypted(body.to_owned(), file), {
+                    info: Some(Box::new(audio_info))
                 });
-                MessageType::Audio(content)
+
+                if let Some(AttachmentInfo::Voice(audio_info, Some(waveform))) = info {
+                    if let Some(duration) = audio_info.duration {
+                        let waveform = waveform.iter().map(|v| (*v).into()).collect();
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/encryption/mod.rs:216]: What's the target type? Would it make sense that the vec in `AttachmentInfo::Voice` immediately store the target type instead?

**Review by bnjbvr — CHANGES_REQUESTED** [review]
woooo voice messages \o/ Gen Z thanks you for your duty.

File: crates/matrix-sdk/src/attachment.rs [hunk 4]
```diff
@@ -87,6 +87,8 @@ pub enum AttachmentInfo {
     Audio(BaseAudioInfo),
     /// The metadata of a file.
     File(BaseFileInfo),
+    /// The metadata of a voice message
+    Voice(BaseAudioInfo, Option<Vec<u16>>),
```
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/attachment.rs]: Done -> https://github.com/matrix-org/matrix-rust-sdk/pull/2697/commits/527f7d71d4151882c1d2636fe54c7d564ca34d98

File: crates/matrix-sdk/src/attachment.rs [hunk 5]
```diff
@@ -77,7 +77,7 @@ pub struct BaseFileInfo {
 }
 
 /// Types of metadata for an attachment.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/attachment.rs]: Applied -> https://github.com/matrix-org/matrix-rust-sdk/pull/2697/commits/0fac318e63df5d6e2998db98e78cd2d59082dd16

File: crates/matrix-sdk/src/media.rs [hunk 6]
```diff
@@ -438,13 +439,23 @@ impl Media {
                 )
             }
             mime::AUDIO => {
-                let info = assign!(info.map(AudioInfo::from).unwrap_or_default(), {
+                let audio_info = assign!(info.clone().map(AudioInfo::from).unwrap_or_default(), {
```
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/media.rs]: Done here -> https://github.com/matrix-org/matrix-rust-sdk/pull/2697/commits/43c59ad0684a4e0c767abe3f9efb194178871658

File: crates/matrix-sdk/src/encryption/mod.rs (hunk 3 above)
Comment (alfogrillo) [reply crates/matrix-sdk/src/encryption/mod.rs]: The final result is a `AudioMessageEventContent`. I don't think it makes sense for `::Voice` to store this ruma type into it.

**Review by bnjbvr — APPROVED** [review]
Thanks!

File: crates/matrix-sdk/src/media.rs [hunk 7]
```diff
@@ -495,3 +492,21 @@ impl Media {
         }
     }
 }
+
+pub(crate) fn update(
```
[the remark this answers is not in the document]
Comment (alfogrillo) [reply crates/matrix-sdk/src/media.rs]: done!

File: crates/matrix-sdk/src/encryption/mod.rs (hunk 3 above)
Comment (alfogrillo) [reply crates/matrix-sdk/src/encryption/mod.rs]: discussed privately

**Pull request merged**